### PR TITLE
[jsk_pcl_ros] extract only directed region of mask image

### PR DESCRIPTION
**Pull request by h-kamada**
* use image_view2 interaction_mode:grabcut_rect
* When use_mask_region, only mask image inside grabcut_rect area is extracted.

### CODE REVIEW

File: jsk_pcl_ros/include/jsk_pcl_ros/mask_image_to_depth_considered_mask_image.h [hunk 1]
```diff
@@ -88,6 +89,14 @@ namespace jsk_pcl_ros
     message_filters::Subscriber<sensor_msgs::PointCloud2> sub_input_;
     message_filters::Subscriber<sensor_msgs::Image> sub_image_;
     ros::Publisher pub_;
+    ros::NodeHandle n_;
+    ros::Subscriber sub_;
+    int region_width;
```
Comment (garaemon) [inline jsk_pcl_ros/include/jsk_pcl_ros/mask_image_to_depth_considered_mask_image.h]: _ suffix

File: jsk_pcl_ros/src/mask_image_to_depth_considered_mask_image_nodelet.cpp [hunk 2]
```diff
@@ -45,19 +46,54 @@ namespace jsk_pcl_ros
...
     pnh_->param("approximate_sync", approximate_sync_, false);
     pub_ = advertise<sensor_msgs::Image>(*pnh_, "output", 1);
-    //    pnh_->param("extract_num", extract_num, 500);
     srv_ = boost::make_shared <dynamic_reconfigure::Server<Config> > (*pnh_);
     dynamic_reconfigure::Server<Config>::CallbackType f =
       boost::bind (
         &MaskImageToDepthConsideredMaskImage::configCallback, this, _1, _2);
     srv_->setCallback (f);
+    sub_ = pnh_->subscribe("input/maskregion", 1, &MaskImageToDepthConsideredMaskImage::mask_region_callback, this);
+    region_width=0;
```
Comment (garaemon) [inline jsk_pcl_ros/src/mask_image_to_depth_considered_mask_image_nodelet.cpp]: spaces between =

File: jsk_pcl_ros/src/mask_image_to_depth_considered_mask_image_nodelet.cpp [hunk 3]
```diff
@@ -45,19 +46,54 @@ namespace jsk_pcl_ros
...
     extract_num_=config.extract_num;
+    use_mask_region=config.use_mask_region;
   }
 
+
+  void MaskImageToDepthConsideredMaskImage::mask_region_callback(const sensor_msgs::Image::ConstPtr& msg){
+    cv_bridge::CvImagePtr cv_ptr = cv_bridge::toCvCopy
+      (msg, sensor_msgs::image_encodings::MONO8);
+    cv::Mat mask = cv_ptr->image;
+    int tmp_width=0;
```
Comment (garaemon) [inline jsk_pcl_ros/src/mask_image_to_depth_considered_mask_image_nodelet.cpp]: spaces between =

File: jsk_pcl_ros/src/mask_image_to_depth_considered_mask_image_nodelet.cpp [hunk 4]
```diff
@@ -86,54 +122,89 @@ namespace jsk_pcl_ros
      const sensor_msgs::PointCloud2::ConstPtr& point_cloud2_msg,
      const sensor_msgs::Image::ConstPtr& image_msg)
   {
-    vital_checker_->poke();
-    pcl::PointCloud<pcl::PointXYZ>::Ptr
-      cloud (new pcl::PointCloud<pcl::PointXYZ>);
-    pcl::fromROSMsg(*point_cloud2_msg, *cloud);
-    pcl::PointCloud<pcl::PointXYZ>::Ptr
-      edge_cloud (new pcl::PointCloud<pcl::PointXYZ>);
+    if(point_cloud2_msg->width==image_msg->width&&point_cloud2_msg->height==image_msg->height){
```
Comment (garaemon) [inline jsk_pcl_ros/src/mask_image_to_depth_considered_mask_image_nodelet.cpp]: space after if

File: jsk_pcl_ros/src/mask_image_to_depth_considered_mask_image_nodelet.cpp [hunk 5]
```diff
@@ -86,54 +122,89 @@ namespace jsk_pcl_ros
...
+            }
+            else{
+              edge_cloud->points[j * width + i] = nan_point;
+            }
+          }
         }
-        else{
-          edge_cloud->points[j * width + i] = nan_point;
+      }
+      else{
```
Comment (garaemon) [inline jsk_pcl_ros/src/mask_image_to_depth_considered_mask_image_nodelet.cpp]: space after else

File: jsk_pcl_ros/src/mask_image_to_depth_considered_mask_image_nodelet.cpp [hunk 6]
```diff
@@ -86,54 +122,89 @@ namespace jsk_pcl_ros
...
+        pcl::KdTreeFLANN<pcl::PointXYZ> kdtree;
+        kdtree.setInputCloud(edge_cloud);
+        pcl::PointXYZ zero;
+        std::vector<int> near_indices;
+        std::vector<float> near_distances;
+        kdtree.nearestKSearch(zero, extract_num_, near_indices, near_distances);
+        ROS_INFO("directed num of extract points:%d   num of nearestKSearch points:%d", extract_num_, ((int) near_indices.size()));
+        int ext_num=std::min(extract_num_, ((int) near_indices.size()));
+        for(int idx=0; idx< ext_num; idx++){
+          int x=near_indices.at(idx)%width;
```
Comment (garaemon) [inline jsk_pcl_ros/src/mask_image_to_depth_considered_mask_image_nodelet.cpp]: spaces between %

**Comment by h-kamada** [comment]
fixed.

File: jsk_pcl_ros/src/mask_image_to_depth_considered_mask_image_nodelet.cpp [hunk 7]
```diff
@@ -45,19 +46,54 @@ namespace jsk_pcl_ros
...
   void MaskImageToDepthConsideredMaskImage::configCallback(Config &config, uint32_t level){
     boost::mutex::scoped_lock lock(mutex_);
-    extract_num_=config.extract_num;
+    extract_num_ = config.extract_num;
+    use_mask_region_ = config.use_mask_region;
+  }
+
+
+  void MaskImageToDepthConsideredMaskImage::mask_region_callback(const sensor_msgs::Image::ConstPtr& msg){
+    cv_bridge::CvImagePtr cv_ptr = cv_bridge::toCvCopy
```
Comment (garaemon) [inline jsk_pcl_ros/src/mask_image_to_depth_considered_mask_image_nodelet.cpp]: You need lock here

File: jsk_pcl_ros/src/mask_image_to_depth_considered_mask_image_nodelet.cpp [hunk 8]
```diff
@@ -86,54 +122,88 @@ namespace jsk_pcl_ros
      const sensor_msgs::PointCloud2::ConstPtr& point_cloud2_msg,
      const sensor_msgs::Image::ConstPtr& image_msg)
   {
-    vital_checker_->poke();
-    pcl::PointCloud<pcl::PointXYZ>::Ptr
-      cloud (new pcl::PointCloud<pcl::PointXYZ>);
-    pcl::fromROSMsg(*point_cloud2_msg, *cloud);
-    pcl::PointCloud<pcl::PointXYZ>::Ptr
-      edge_cloud (new pcl::PointCloud<pcl::PointXYZ>);
+    if (point_cloud2_msg->width == image_msg->width && point_cloud2_msg->height == image_msg->height){
```
Comment (garaemon) [inline jsk_pcl_ros/src/mask_image_to_depth_considered_mask_image_nodelet.cpp]: You need lock here

**Comment by garaemon** [comment]
You need lock like https://github.com/jsk-ros-pkg/jsk_recognition/blob/master/jsk_pcl_ros/src/hinted_stick_finder_nodelet.cpp#L99

**Comment by h-kamada** [comment]
fixed

**Pull request merged**